include/ck/wrapper/tensor.hpp

```diff
@@ -10,324 +10,413 @@
+     */
+    template <typename... Ts, enable_if_t<!detail::HasSlice(Tuple<Ts...>{}), bool> = false>
     __host__ __device__ ElementType& operator[](const Tuple<Ts...>& idx)
     {
         if constexpr(IsDynamicBuffer)
         {
-            const index_t offset = layout_(idx);
+            const index_t offset = layout_(idx) + base_offset_;
             return buffer_(offset);
         }
         else
         {
-            constexpr index_t offset = Layout<Shape, UnnestedDescriptorType>{
+            constexpr index_t index_offset = Layout<Shape, UnrolledDescriptorType>{
+                Shape{},
+                UnrolledDescriptorType{}}.template operator()<Tuple<Ts...>>();
+            // Apply embed offset (calculate in compiletime)
+            constexpr index_t base_offset = Layout<Shape, UnrolledDescriptorType>{
                 Shape{},
-                UnnestedDescriptorType{}}.template operator()<Tuple<Ts...>>();
-            return buffer_(Number<offset>{});
+                UnrolledDescriptorType{}}.template operator()<MultiIndex<Shape::Size()>>();
+            return buffer_(Number<index_offset + base_offset>{});
         }
     }
 
-    template <typename... Ts, enable_if_t<!IsSlicing(Tuple<Ts...>{}), bool> = false>
+    template <typename... Ts, enable_if_t<!detail::HasSlice(Tuple<Ts...>{}), bool> = false>
     __host__ __device__ ElementType& operator()(const Tuple<Ts...>& idx)
     {
         return this->operator[](idx);
     }
 
-    template <typename... Idxs, enable_if_t<!IsSlicing(Tuple<Idxs...>{}), bool> = false>
+    template <typename... Idxs, enable_if_t<!detail::HasSlice(Tuple<Idxs...>{}), bool> = false>
     __host__ __device__ ElementType& operator()(Idxs... idxs)
     {
         return this->operator[](make_tuple(idxs...));
     }
 
-    __host__ __device__ constexpr auto GetDefaultDescriptor()
+    /**
+     * \brief Get default layout descriptor.
+     *
+     * \return Default layout descriptor.
+     */
+    __host__ __device__ constexpr auto GetMergedNestsDescriptor()
     {
-        return layout_.GetDefaultDescriptor();
+        return layout_.GetMergedNestsDescriptor();
     }
 
+    /**
+     * \brief Get pointer to the data.
+     *
+     * \return Pointer.
+     */
     __host__ __device__ ElementType* GetPointer() const { return buffer_.p_data_; }
 
+    __host__ __device__ constexpr auto& GetBuffer() { return buffer_; }
+    __host__ __device__ constexpr auto& GetBuffer() const { return buffer_; }
+
+    /**
+     * \brief Get multi index offset to the data.
+     *
+     * \return Multi index offset.
+     */
+    __host__ __device__ constexpr auto& GetMultiIdxOffsets() const { return multi_idx_offset_; }
+
+    /**
+     * \brief Apply multi index offset on the tensor.
+     *
+     * \param multi_idx_offset Multi index offset.
+     */
+    template <typename MultiIdxOffsets>
+    __host__ __device__ constexpr void ShiftMultiIdxOffset(const MultiIdxOffsets multi_idx_offset)
```

Comment:
This should rather be renamed to `SetMultiIdxOffset` because you're not doing any arithmetic operation on it, but just rather setting it's value.